x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/PlannerUtils.java

```diff
@@ -73,6 +82,35 @@ public static Tuple<PhysicalPlan, PhysicalPlan> breakPlanBetweenCoordinatorAndDa
         return new Tuple<>(coordinatorPlan, dataNodePlan.get());
     }
 
+    public static PhysicalPlan dataNodeReductionPlan(LogicalPlan plan, PhysicalPlan unused) {
+        var pipelineBreakers = plan.collectFirstChildren(Mapper::isPipelineBreaker);
+
+        if (pipelineBreakers.isEmpty() == false) {
+            UnaryPlan pipelineBreaker = (UnaryPlan) pipelineBreakers.get(0);
+            if (pipelineBreaker instanceof TopN topN) {
```

Comment:
I'd love it if we had an interface or something we could slap on these and let them implement a method to do this replacement. Or a method on `LogicalPlan`. I dunno. As ESQL is going to get touched by more people we need to have the abstractions be readable. I know we don't want to mix optimization stuff into the models, but we're really doing that here already, it's just that the code isn't in the place where Java developers will look for it.